Dev

tigunov-it opened the pull request:
Queryparams
13. does your solution supports passing queryparams into angular?
https://github.com/tigunov-it/angular-react-starter-test/blob/637e2ec808bdc9f8df9d0234605702f064058854/nginx.conf#L39

    -- Answer: That worked for me on Angular

14.  Is it good to use port 80 inside container?
https://github.com/tigunov-it/angular-react-starter-test/blob/637e2ec808bdc9f8df9d0234605702f064058854/nginx.conf#L31


  -- Answer: The official nginx docker image on dockerhub is set to port 80 inside the container by default. Also in the official nginx documentation in the examples port 80. I think this is the best solution, but it's interesting to know your opinion.

dmitry-mightydevops commented:
@tigunov-it 
> `try_files $uri $uri/ =404;`  so that doesn't pass query params, it probably has worked for you because you don't have any functionality in test angular app that is using it, but you're not passing it and therefore it doesn't work. Pls adjust the configuration. See nginx docs. So for example you should pass url/?hello=world and that "hello" variable inside angular should be set properly.

> port 80
https://www.w3.org/Daemon/User/Installation/PrivilegedPorts.html
It's never a good idea to use such a low number as it requires elevated privileges for the container, especially if you run it inside kubernetes.

dmitry-mightydevops reviewed (changes requested):
Yuri, this PR does contain only email change. It does not contain all the requested changes:

---
Good start! Few comments:

1. NPM Build inside dockerfile is not good way of getting things done
It should be done outside the container and then you operate on pure files inside dockerfile
Please find a way to get it done in the ci.yaml.

2. Sending emails clogs the code. A lot of copy paste.
Please optimize your code so it's much shorter. 

Hint:
https://docs.github.com/en/actions/creating-actions/creating-a-composite-action


3. Comments

Thank you for Comments
However they're hard to read and they double the amount of horizontal lines. 
Put the comments (aligned) after the actual yaml element:

i.e:
```
this:
  is: 1    # comment
  and: 2   # comment 2
  more: 3  # comment 3
```

4. No hardcode values
 to: 192215@gmail.com

 What if a different author commits the change?

5. cancel step
No need to send email if the deployment was cancelled

6. get-job-time:

Authenticate using GITHUB_TOKEN
https://docs.github.com/en/actions/security-guides/automatic-token-authentication

6.1 get-job-time should be called prior to sending success email.

I.e. the workflow
- build complete successfully
- you measure execution time
- you send succesful email with timing info included in body. No need to attach JSON file - this is not user friendly. So parse API response prior to sending the email and form a nice email's body.

No need for a bunch of atomic emails. 

1. Use tags not sha1 tags
uses: docker/login-action@f054a8b539a109f9f41c372932f1ae047eff08c9

i.e. latest, v1 etc. SHA makes it hard for humans.

8. success/fail

Find a better way to send mail. This takes entire screen

```
### Sending email when deploy successfull
      - name: Send mail
        if: ${{ success() }}
        uses: dawidd6/action-send-mail@v3
        with:
          # Required mail server address:
          server_address: smtp.gmail.com
          # Required mail server port:
          server_port: 465
          # Optional (recommended): mail server username:
          username: ${{secrets.MAIL_USERNAME}}
          # Optional (recommended) mail server password:
          password: ${{secrets.MAIL_PASSWORD}}
          # Required mail subject:
          subject: Github Actions job result
          # Required recipients' addresses:
          to: 192215@gmail.com
          # Required sender full name (address can be skipped):
          from: tigunov-it # <user@example.com>
          # Optional whether this connection use TLS (default is true if server_port is 465)
          secure: true
          # Optional plain body:
          body: Build job of ${{github.repository}} completed successfully!
          # Optional HTML body read from file:
          #html_body: file://README.html
          # Optional carbon copy recipients:
          #cc: kyloren@example.com,leia@example.com
          # Optional blind carbon copy recipients:
          # bcc: r2d2@example.com,hansolo@example.com
          # Optional recipient of the email response:
          reply_to: 192215@gmail.com
          # Optional Message ID this message is replying to:
          # in_reply_to: <random-luke@example.com>
          # Optional unsigned/invalid certificates allowance:
          ignore_cert: true
          # Optional converting Markdown to HTML (set content_type to text/html too):
          convert_markdown: true
          # Optional attachments:
          # attachments: attachments.zip,git.diff,./dist/static/*.js
          # Optional priority: 'high', 'normal' (default) or 'low'
          priority: normal

      ### Sending email when deploy failed
      - name: Send mail
        if: ${{ failure() }}
        uses: dawidd6/action-send-mail@v3
        with:
          # Required mail server address:
          server_address: smtp.gmail.com
          # Required mail server port:
          server_port: 465
          # Optional (recommended): mail server username:
          username: ${{secrets.MAIL_USERNAME}}
          # Optional (recommended) mail server password:
          password: ${{secrets.MAIL_PASSWORD}}
          # Required mail subject:
          subject: Github Actions job result
          # Required recipients' addresses:
          to: 192215@gmail.com
          # Required sender full name (address can be skipped):
          from: tigunov-it # <user@example.com>
          # Optional whether this connection use TLS (default is true if server_port is 465)
          secure: true
          # Optional plain body:
          body: Build job of ${{github.repository}} is failed!
          # Optional HTML body read from file:
          #html_body: file://README.html
          # Optional carbon copy recipients:
          #cc: kyloren@example.com,leia@example.com
          # Optional blind carbon copy recipients:
          # bcc: r2d2@example.com,hansolo@example.com
          # Optional recipient of the email response:
          reply_to: 192215@gmail.com
          # Optional Message ID this message is replying to:
          # in_reply_to: <random-luke@example.com>
          # Optional unsigned/invalid certificates allowance:
          ignore_cert: true
          # Optional converting Markdown to HTML (set content_type to text/html too):
          convert_markdown: true
          # Optional attachments:
          # attachments: attachments.zip,git.diff,./dist/static/*.js
          # Optional priority: 'high', 'normal' (default) or 'low'
          priority: normal
```


9. When you done the changes - pls open the PR and assign me as reviewer. My github username is dmitry-mightydevops

10. Pls also join me as collaborator to the repo.

11. Your nodes versions mismatch in dockerfile and ci.yaml
https://github.com/tigunov-it/angular-react-starter-test/blob/637e2ec808bdc9f8df9d0234605702f064058854/Dockerfile#L2
You don't need NPM buld stage 1 as it should be done in ci.yaml

11.1 prod env

also keep in mind that you should not hardcode env name
RUN npm run build --prod

the same ci file should be used to build prod/staging/dev envs. Therefore use the name branch to calculate and then pass a proper argument into npm build.

12. In dockerfile always stick to either SHA1 revs or named tags. So we know what nginx version you're using
https://github.com/tigunov-it/angular-react-starter-test/blob/637e2ec808bdc9f8df9d0234605702f064058854/Dockerfile#L10


13. Queryparams

does your solution supports passing queryparams into angular?
https://github.com/tigunov-it/angular-react-starter-test/blob/637e2ec808bdc9f8df9d0234605702f064058854/nginx.conf#L39

14. Is it good to use port 80 inside container?
https://github.com/tigunov-it/angular-react-starter-test/blob/637e2ec808bdc9f8df9d0234605702f064058854/nginx.conf#L31

tigunov-it commented:
> Yuri, this PR does contain only email change. It does not contain all the requested changes:
According to all the comments, I did the work. Initially, I used one branch of GIT (master), so I uploaded all the changes to it.

>     1. NPM Build inside dockerfile is not good way of getting things done
>        It should be done outside the container and then you operate on pure files inside dockerfile
>        Please find a way to get it done in the ci.yaml.
Previously, I used the assembly inside the dockerfile, now the assembly takes place in the github action (ci.yaml)

> 
>     2. Sending emails clogs the code. A lot of copy paste.
>        Please optimize your code so it's much shorter.
> Hint: https://docs.github.com/en/actions/creating-actions/creating-a-composite-action
I followed your suggestion and explored the possibility of using Composite Action and applied this solution.

>     3. Comments
> Thank you for Comments However they're hard to read and they double the amount of horizontal lines. Put the comments (aligned) after the actual yaml element:
For better readability of the code, I moved the comments to a horizontal line.

>     4. No hardcode values
>        to: [192215@gmail.com](mailto:192215@gmail.com)
To avoid hard coding, I moved the email to a variable.
> 
>     5. cancel step
>        No need to send email if the deployment was cancelled
Sending email when a build is canceled is excluded.

>     6. get-job-time:
> Authenticate using GITHUB_TOKEN https://docs.github.com/en/actions/security-guides/automatic-token-authentication
I used GITHUB_TOKEN when requesting the API information about runner
> 
> 6.1 get-job-time should be called prior to sending success email.
> 
> I.e. the workflow
>     * build complete successfully
>     * you measure execution time
>     * you send succesful email with timing info included in body. No need to attach JSON file - this is not user friendly. So parse API response prior to sending the email and form a nice email's body.
> No need for a bunch of atomic emails.
After requesting the build information API, I cut the information from the json file and paste it into the body of the email:
      - run: curl --user tigunov-it:${{ secrets.GITHUB_TOKEN }} https://api.github.com/repos/tigunov-it/angular-react-starter-test/actions/runs/${GITHUB_RUN_ID}/timing  > time.txt
      - name: Parse job time
        run: echo "::set-output name=JOB_TIME::$(cat time.txt | grep "duration_ms")"
        id: time
>     1. Use tags not sha1 tags
>        uses: [docker/login-action@f054a8b](https://github.com/docker/login-action/commit/f054a8b539a109f9f41c372932f1ae047eff08c9)
> i.e. latest, v1 etc. SHA makes it hard for humans.
I removed cha1 and used tags. (docker/login-action@v2)

>     8. success/fail
> 
> Find a better way to send mail. This takes entire screen
> ```
I moved the handling of sending mail to the Composite Action, and also moved the comments to a horizontal line - this significantly reduced the code.
>     9. When you done the changes - pls open the PR and assign me as reviewer. My github username is dmitry-mightydevops
Done
> 
>     10. Pls also join me as collaborator to the repo.
Done
> 
>     11. Your nodes versions mismatch in dockerfile and ci.yaml
>         https://github.com/tigunov-it/angular-react-starter-test/blob/637e2ec808bdc9f8df9d0234605702f064058854/Dockerfile#L2
fixed
      
>         You don't need NPM buld stage 1 as it should be done in ci.yaml
fixed

> 11.1 prod env
> also keep in mind that you should not hardcode env name RUN npm run build --prod
> 
> the same ci file should be used to build prod/staging/dev envs. Therefore use the name branch to calculate and then pass a proper argument into npm build.
I use if: ${{ steps.extract_branch.outputs.branch == 'master' }} # if branch = master, then npm run build --prod
Is it a good idea to build depending on the branch of the repository?

>     12. In dockerfile always stick to either SHA1 revs or named tags. So we know what nginx version you're using
>         https://github.com/tigunov-it/angular-react-starter-test/blob/637e2ec808bdc9f8df9d0234605702f064058854/Dockerfile#L10
fixed
> 
>     13. Queryparams
> does your solution supports passing queryparams into angular?
> 
> https://github.com/tigunov-it/angular-react-starter-test/blob/637e2ec808bdc9f8df9d0234605702f064058854/nginx.conf#L39
I changed to
try_files $uri $uri/ /index.html;

>     14. Is it good to use port 80 inside container?
>         https://github.com/tigunov-it/angular-react-starter-test/blob/637e2ec808bdc9f8df9d0234605702f064058854/nginx.conf#L31
I have read about privileged ports. Thank you.
I realized that it's better to change the port inside the container.

dmitry-mightydevops commented on docker-compose.yaml:
```diff
@@ -3,4 +3,4 @@ services:
     image: ghcr.io/tigunov-it/angular-react-starter-test:master
     ports:
-      - "8080:80"
+      - "8080:8080"
```
+1!

dmitry-mightydevops commented on nginx.conf:
```diff
@@ -28,15 +28,18 @@ http {
     location / {
-      try_files $uri $uri/ =404;
+#       try_files $uri $uri/ =404;
+        try_files $uri $uri/ /index.html;
```
so this is not exactly what I was looking for. Please read about $args nginx variable:
http://nginx.org/en/docs/http/ngx_http_core_module.html#var_args
or ($query_string)

tigunov-it replied:
Maybe we can use this solution?
try_files $uri$args $uri$args/ $uri $uri/ /index.html =404;

dmitry-mightydevops replied:
yes that is better! :)

dmitry-mightydevops reviewed (changes requested):
1. https://github.com/tigunov-it/angular-react-starter-test/blob/dcae0917aeceacc15f8eedbeb699e0132ff6bf08/.github/workflows/ci.yaml#L21
this is available in ENV vars that githup runners sets for you
see: https://docs.github.com/en/actions/learn-github-actions/environment-variables

2. not sure I understand why you make the check here
```
- name: Use Node.js 16
  if: ${{ steps.extrac
  uses: actions/setup-
  with:
    node-version: 16.x
```

because you always trigger the action only by pushing in the master branch. 
```
on:
  push:
    branches: [ master ]
```
as a result that IF condition is not necessary since it provides no value at all


2.1 technically speaking you need to install node every time the action runs
```
uses: actions/setup-node@v3
```

2.2 it's better to add additional branch name like feature/ci* that you use to debug your github actions without pushing anything to production, that way you can keep working on the changes to your GHA without distutrbing live envs or developers doing the work
Like I mentioned earlier - each run requires the installation of the nodejs.

4. Please remove steps that do nothing (debug steps). 
Make clean pipelines

```
- name: test directory
  run: ls -la

- name: Show message body
  run: echo ${{ env.message_body }} 
``` 

5. Your job should output the value that should not require any parsing in the next step. Use jq to get the value from the API json output
```
- name: get-job-time
  run: curl --user tigunov-it:${{ secrets.GITHUB_TOKEN }} https://api.github.com/repos/tigunov-it/angular-react-starter-test/actions/runs/${GITHUB_RUN_ID}/timing  > time.txt

- name: Parse job time
  run: echo "::set-output name=JOB_TIME::$(cat time.txt | grep "duration_ms")"
  id: time
```

dmitry-mightydevops reviewed (changes requested):
pls also clean up comments and "debugs" from other files 
like https://github.com/tigunov-it/angular-react-starter-test/blob/dcae0917aeceacc15f8eedbeb699e0132ff6bf08/nginx.conf#L31

Make sure your code is clean, formatted, passed linters and other engineers would love to jump and support your solution going forward!

tigunov-it commented:
> pls also clean up comments and "debugs" from other files like
> 
> https://github.com/tigunov-it/angular-react-starter-test/blob/dcae0917aeceacc15f8eedbeb699e0132ff6bf08/nginx.conf#L31
> 
> Make sure your code is clean, formatted, passed linters and other engineers would love to jump and support your solution going forward!

Done. I delete dubug code.

tigunov-it commented:
> 1. https://github.com/tigunov-it/angular-react-starter-test/blob/dcae0917aeceacc15f8eedbeb699e0132ff6bf08/.github/workflows/ci.yaml#L21
>        
>        this is available in ENV vars that githup runners sets for you
>        see: https://docs.github.com/en/actions/learn-github-actions/environment-variables
> 
>     2. not sure I understand why you make the check here
> ```
> - name: Use Node.js 16
>   if: ${{ steps.extrac
>   uses: actions/setup-
>   with:
>     node-version: 16.x
> ```
> 
> because you always trigger the action only by pushing in the master branch.
> 
> ```
> on:
>   push:
>     branches: [ master ]
> ```
> 
> as a result that IF condition is not necessary since it provides no value at all

I changed the trigger. Now you can get the name of the branch and, depending on this, execute certain build and test commands.

> 
> 2.1 technically speaking you need to install node every time the action runs
> ```
> uses: actions/setup-node@v3
> ```
I didn't understand this question. What goal do we want to achieve?

> 
> 2.2 it's better to add additional branch name like feature/ci* that you use to debug your github actions without pushing anything to production, that way you can keep working on the changes to your GHA without distutrbing live envs or developers doing the work Like I mentioned earlier - each run requires the installation of the nodejs.
> 
It is not clear to me how to test the assembly of a package on a separate branch. My test suite will still get in the way. At the moment I have created a second repository for testing and experimentation.

>     4. Please remove steps that do nothing (debug steps).
>        Make clean pipelines
Done

> 
>     5. Your job should output the value that should not require any parsing in the next step. Use jq to get the value from the API json output
> 
Do you mean use JQ linux command? I went to test and study this command.
I found that the api request only gives correct runtime information on private repositories. Most likely this is information about the use of paid time. On the public repository, time = 0.

Perhaps a simpler solution would be to use comand uptime? I am testing this solution.

tigunov-it commented:
>     5. Your job should output the value that should not require any parsing in the next step. Use jq to get the value from the API json output
> ```
> - name: get-job-time
>   run: curl --user tigunov-it:${{ secrets.GITHUB_TOKEN }} https://api.github.com/repos/tigunov-it/angular-react-starter-test/actions/runs/${GITHUB_RUN_ID}/timing  > time.txt
> 
> - name: Parse job time
>   run: echo "::set-output name=JOB_TIME::$(cat time.txt | grep "duration_ms")"
>   id: time
> ```

dmitry-mightydevops commented:
> Do you mean use JQ linux command? I went to test and study this command.
I found that the api request only gives correct runtime information on private repositories. Most likely this is information about the use of paid time. On the public repository, time = 0.
Perhaps a simpler solution would be to use comand uptime? I am testing this solution.

If API doesn't give you any information - then you have options
- use existing action that tracks execution time (I know at least one exists)
- store the start time in task results and consume it (comparing with current time at the end) before sending the email.

dmitry-mightydevops commented on .github/workflows/ci.yaml:
```diff
@@ -18,16 +16,8 @@ jobs:
-
       - name: Use Node.js 16.x
-        if: ${{ steps.extract_branch.outputs.branch == 'master' }}
+        if: ${{ github.ref }} == 'master'
```
+1!

dmitry-mightydevops commented on .github/workflows/ci.yaml:
```diff
@@ -92,11 +79,8 @@ jobs:
-
       - name: Parse job time
-        run: echo "::set-output name=JOB_TIME::$(cat time.txt | grep "duration_ms")"
+        run: echo "::set-output name=JOB_TIME::$(uptime)"
```
so we need to know the exact value in seconds. Like the duration was 72seconds. Which is displayed in the final email.

tigunov-it replied:
I use "date +%s" at the beginning and at the end workflow.
It works correctly

dmitry-mightydevops commented on nginx.conf:
```diff
@@ -38,11 +35,7 @@ http {
     location / {
-#       try_files $uri $uri/ =404;
-        try_files $uri $uri/ /index.html;
+    try_files $uri$args $uri$args/ $uri $uri/ /index.html =404;
```
args - should be used at the final URI
something like this:
```
location / {
  try_files $uri $uri/ /index.php?$args;
}
```

Pls test your solution.

dmitry-mightydevops replied:
Check for various tutorials online for location and query_params parsing.

tigunov-it replied:
location / {
  try_files $uri $uri/ /index.php?$args;
}
This solution works

dmitry-mightydevops replied:
perfect!

dmitry-mightydevops commented on .github/workflows/ci.yaml:
```diff
@@ -1,8 +1,6 @@
-on:
-  push:
-    branches: [ master ]
+on: push
```
so the problem is this:
any branch commit will result in this action triggered. feature/branch that developer is using for current development will push changes to the environment - which is 100% risky and buggy.

you should limit what branches triggers the build.

Also what I meant in my previous feedback is, sometimes you need to test a new version of the pipeline, but you don't need/want to break the current production pipeline.

You start a new branch `feature/ci-improvements` and perform the changes there, while main branch still has your OLD implementation (that is supposed to be working 100% of the time). Then once you are comfortable with your changes - your merge `feature/ci-improvements -> master` and your master has a new pipeline implementation and you have 0% downtime.

So test this scenario and limit the branches you can trigger the github action

tigunov-it replied:
I realized that the build on each branch makes its own package with the appropriate tag. Thus, packages for the production environment, as well as for testing and development, can be stored in the repository. Depending on the branch, we will specify different parameters in ci.yaml:
for master:
on:
  push:
    branches: [ master ]

for dev:
on:
  push:
    branches: [ dev ]
for feature/ci-improvements:
on:
  push:
    branches: [ feature/ci-improvements]
